### PR TITLE
progress towards SQL case processing

### PR DESCRIPTION
**Pull request by snopoke**
@benrudolph so close but not there quite yet

cc @biyeun 

:blowfish: (sorry, lots of commits!)

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'ABCMeta' imported but unused
* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'abstractproperty' imported but unused
* corehq/form_processor/backends/sql/casedb.py, line 48 - E501 line too long (116 > 115 characters)
* corehq/form_processor/models.py, line 42 - E501 line too long (118 > 115 characters)
* corehq/form_processor/models.py, line 306 - E501 line too long (116 > 115 characters)
* corehq/form_processor/update_strategy_base.py, line 19 - E501 line too long (117 > 115 characters)

**Comment by benrudolph** [comment]
awesome, will review after lunch

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'ABCMeta' imported but unused
* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'abstractproperty' imported but unused
* corehq/form_processor/backends/sql/casedb.py, line 48 - E501 line too long (116 > 115 characters)
* corehq/form_processor/models.py, line 42 - E501 line too long (118 > 115 characters)
* corehq/form_processor/models.py, line 306 - E501 line too long (116 > 115 characters)
* corehq/form_processor/update_strategy_base.py, line 19 - E501 line too long (117 > 115 characters)

**Comment by benrudolph** [comment]
damn, nice one. looking good will wait for tests. probably makes sense to address feedback in separate PR

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'ABCMeta' imported but unused
* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'abstractproperty' imported but unused
* corehq/form_processor/backends/sql/casedb.py, line 48 - E501 line too long (116 > 115 characters)
* corehq/form_processor/models.py, line 42 - E501 line too long (118 > 115 characters)
* corehq/form_processor/models.py, line 306 - E501 line too long (116 > 115 characters)

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'ABCMeta' imported but unused
* corehq/form_processor/backends/couch/update_strategy.py, line 1 - F401 'abstractproperty' imported but unused
* corehq/form_processor/backends/sql/casedb.py, line 48 - E501 line too long (116 > 115 characters)
* corehq/form_processor/models.py, line 42 - E501 line too long (118 > 115 characters)
* corehq/form_processor/models.py, line 306 - E501 line too long (116 > 115 characters)

File: corehq/form_processor/models.py [hunk 1]
```diff
@@ -24,7 +24,22 @@
...
+            raise TypeError("Form instances without form ID value are unhashable")
+        return hash(hash_val)
+
+
+class SaveStateMixin(object):
+    def is_saved(self):
+        return bool(self._get_pk_val())
+
+
+class XFormInstanceSQL(PreSaveHashableMixin, models.Model, AbstractXFormInstance, RedisLockableMixIn, SaveStateMixin):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (118 > 115 characters)

File: corehq/form_processor/models.py [hunk 2]
```diff
@@ -277,7 +302,10 @@ class XFormMetadata(jsonobject.JsonObject):
     appVersion = jsonobject.StringProperty()
     location = GeoPointProperty()
 
-class CommCareCaseSQL(models.Model, AbstractCommCareCase, RedisLockableMixIn):
+
+class CommCareCaseSQL(PreSaveHashableMixin, models.Model, AbstractCommCareCase, RedisLockableMixIn, SaveStateMixin):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (116 > 115 characters)

**Comment by benrudolph** [comment]
:white_check_mark: looks good once tag is off

**Comment by snopoke** [comment]
Phew, that was quite a fight!

**Comment by czue** [comment]
you guys are crushing this!

**Comment by benrudolph** [comment]
:hammer: :hammer: :hammer:

**Pull request merged**